ext/drstatecmp/drstatecmp.dox

```diff
@@ -0,0 +1,146 @@
+ *
+ * * Neither the name of Google, Inc. nor the names of its contributors may be
+ *   used to endorse or promote products derived from this software without
+ *   specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL GOOGLE, INC. OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ */
+
+/**
+***************************************************************************
+***************************************************************************
+\page page_drstatecmp Machine State Comparison Library
+
+The \p drstatecmp DynamoRIO Machine State Comparison Extension provides mechanisms
+to enable systematic and exhaustive machine state comparisons across instrumentation
+sequences. These comparisons will detect instrumentation-induced corruptions of the
+application's machine state that would typically lead to obscure bugs.
+
+ - \ref sec_drstatecmp_init
+ - \ref sec_drstatecmp_checks
+ - \ref sec_drstatecmp_instr_state_sharing
+ - \ref sec_drstatecmp_machine_state_saved
+ - \ref sec_drstatecmp_vs_drbbdup
+
+\section sec_drstatecmp_init Setup
+
+To use \p drstatecmp with your client simply include this line in your client's
+\p CMakeLists.txt file:
+
+\code use_DynamoRIO_extension(clientname drstatecmp) \endcode
+
+That will automatically set up the include path and library dependence.
+
+The initialization routine \p drstatecmp_init() must be called prior to any of
+the other routines.
+
+\section sec_drstatecmp_checks Checks
+
+There are two types of checks depending on whether a basic block has any side-effects.
+
+\subsection sec_drstatecmp_checks_side_effects_free Side-effect-free Basic Blocks
+
+Side-effect-free basic blocks are duplicated and executed twice, once with the full
+instrumentation sequence and once with no instrumentation except for a state comparison at
+the end. Essentially, we check whether the machine state at the end of the basic block is
+the same with and without instrumentation code. This check should catch any state
+clobbering by instrumentation.
+
+The side-effect-free basic blocks are duplicated at drmgr's app2app stage. The copy of
+each such basic block is appended immediately after the original basic block. Labels are
+inserted in the beginning of the original basic block (ORIG_BB) and its copy (COPY_BB) to
```

Comment:
Use `\p ORIG_BB` or sthg to indicate it's a literal code term.  Ditto for the others.